ludwig/features/category_feature.py

```diff
@@ -388,61 +393,44 @@ def postprocess_predictions(
-                    metadata['idx2str'][pred] for pred in preds
-                ]
-
-            else:
-                postprocessed[PREDICTIONS] = preds
-
-            if not skip_save_unprocessed_output:
-                np.save(npy_filename.format(name, PREDICTIONS), preds)
-
-            del predictions[PREDICTIONS]
-
-        if PROBABILITIES in predictions and len(
-                predictions[PROBABILITIES]) > 0:
-            probs = predictions[PROBABILITIES].numpy()
-            prob = np.amax(probs, axis=1)
-            postprocessed[PROBABILITIES] = probs
-            postprocessed[PROBABILITY] = prob
-
-            if not skip_save_unprocessed_output:
-                np.save(npy_filename.format(name, PROBABILITIES), probs)
-                np.save(npy_filename.format(name, PROBABILITY), probs)
-
-            del predictions[PROBABILITIES]
-
-        if ('predictions_top_k' in predictions and
-            len(predictions['predictions_top_k'])) > 0:
+                predictions[predictions_col] = backend.df_engine.map_objects(
+                    predictions[predictions_col],
+                    lambda pred: metadata['idx2str'][pred]
+                )
 
-            preds_top_k = predictions['predictions_top_k'].numpy()
+        probabilities_col = f'{self.feature_name}_{PROBABILITIES}'
+        if probabilities_col in predictions:
+            prob_col = f'{self.feature_name}_{PROBABILITY}'
+            predictions[prob_col] = predictions[probabilities_col].map(max)
+            predictions[probabilities_col] = backend.df_engine.map_objects(
+                predictions[probabilities_col],
+                lambda pred: pred.tolist()
+            )
             if 'idx2str' in metadata:
-                postprocessed['predictions_top_k'] = [
-                    [metadata['idx2str'][pred] for pred in pred_top_k]
-                    for pred_top_k in preds_top_k
-                ]
-            else:
-                postprocessed['predictions_top_k'] = preds_top_k
+                for i, label in enumerate(metadata['idx2str']):
+                    key = f'{probabilities_col}_{label}'
+
+                    # Use default param to force a capture before the loop completes, see:
+                    # https://stackoverflow.com/questions/2295290/what-do-lambda-function-closures-capture
+                    predictions[key] = backend.df_engine.map_objects(
+                        predictions[probabilities_col],
+                        lambda prob, i=i: prob[i],
+                    )
 
-            if not skip_save_unprocessed_output:
-                np.save(
-                    npy_filename.format(name, 'predictions_top_k'),
-                    preds_top_k
+        top_k_col = f'{self.feature_name}_predictions_top_k'
+        if top_k_col in predictions:
+            if 'idx2str' in metadata:
+                predictions[top_k_col] = backend.df_engine.map_objects(
+                    predictions[top_k_col],
```

Comment:
not 100% sure of how the content of the top k column looks like, but before each was a list of values, so in line 431 there was a loop for each value, while here it sems like there's only a loop over preds and not inside each pred

Reply:
Good catch, every row here should be a `pred_top_k`.  Fixed.